# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: 

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [324]:
import numpy as np
import pandas as pd
import yellowbrick


### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [325]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y
X, y = yellowbrick.datasets.loaders.load_spam(data_home=None, return_dataset=False)
print(f"Size of X is: {X.size}, Type of X is: {type(X)}")
print(f"Size of y is: {y.size}, Type of y is: {type(y)}")


Size of X is: 262200, Type of X is: <class 'pandas.core.frame.DataFrame'>
Size of y is: 4600, Type of y is: <class 'pandas.core.series.Series'>


### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [326]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_value = X.isnull().sum().sum()+y.isnull().sum()
print(f"There are {missing_value} missing values in the dataset")

There are 0 missing values in the dataset


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [327]:
# TO DO: Create X_small and y_small 
from sklearn.model_selection import train_test_split
_, X_small, _, y_small = train_test_split(X, y, test_size=0.05, random_state= 0)
# print(X_small_test)

### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [328]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=2000)
#Train_test_split and fit with training data
#First machine learning model
X_train,X_val,y_train,y_val = train_test_split(X,y, random_state= 0)
model1 = model.fit(X_train, y_train)

#Second machine learning model
X_col = X.iloc[:,:2]
X_col_train, X_col_val, y_train, y_val = train_test_split(X_col, y, random_state= 0)
model2 = model.fit(X_col_train, y_train)

#Third machine learning model
X_small_train, X_small_val, y_small_train, y_small_val = train_test_split(X_small, y_small, random_state= 0)
model3 = model.fit(X_small_train, y_small_train)




### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

In [329]:
X_train,X_val,y_train,y_val = train_test_split(X,y, random_state= 0)
model1 = model.fit(X_train, y_train)
model1_train_score = model1.score(X_train, y_train)
model1_test_score = model1.score(X_val, y_val)

print("Model 1 Training set score: {:.2f}".format(model1.score(X_train, y_train)))
print("Model 1 Validation set score: {:.2f}".format(model1.score(X_val, y_val)))

X_col = X.iloc[:,:2]
X_col_train, X_col_val, y_train, y_val = train_test_split(X_col, y, random_state= 0)
model2 = model.fit(X_col_train, y_train)
model2_train_score = model2.score(X_col_train, y_train)
model2_test_score = model2.score(X_col_val, y_val)

print("Model 2 Training set score: {:.2f}".format(model2.score(X_col_train, y_train)))
print("Model 2 Validation set score: {:.2f}".format(model.score(X_col_val, y_val)))

X_small_train, X_small_val, y_small_train, y_small_val = train_test_split(X_small, y_small, random_state= 0)
model3 = model.fit(X_small_train, y_small_train)
model3_train_score = model3.score(X_small_train, y_small_train)
model3_test_score = model3.score(X_small_val, y_small_val)

print("Model 3 Training set score: {:.2f}".format(model.score(X_small_train, y_small_train)))
print("Model 3 Validation set score: {:.2f}".format(model.score(X_small_val, y_small_val)))

Model 1 Training set score: 0.93
Model 1 Validation set score: 0.94
Model 2 Training set score: 0.61
Model 2 Validation set score: 0.61
Model 3 Training set score: 0.97
Model 3 Validation set score: 0.79


### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [330]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

results = pd.DataFrame(columns = ['Data Size','Training Accuracy','Validation Accuracy'])
results.loc[0] = [X.shape, model1_train_score, model1_test_score]
results.loc[1] = [X_col.shape, model2_train_score, model2_test_score]
results.loc[2] = [X_small.shape, model3_train_score, model3_test_score]
results.index = ['Full DataSet', 'First two columns', 'Smaller Dataset']

print(results)

                    Data Size  Training Accuracy  Validation Accuracy
Full DataSet       (4600, 57)           0.928696             0.936522
First two columns   (4600, 2)           0.608406             0.613043
Smaller Dataset     (230, 57)           0.965116             0.793103


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*
1. When we have the full dataset available, we can see that the training score and accuracy score are both pretty high and close to each other, with the a difference of less than 1%. 
When we only use the first two columns, our training score drops significantly versus the full dataset. This means using two features is not enough information for our model to make accurate predictions.
And the last scenario with a much smaller dataset shows a good score, but the validation score is much lower meaning our model is overfitting.
2. In this case, for the spam dataset, a false positive would be when the model incorrectly identifies a regular email as a spam email. Which would place an important email in the spam folder.
On the other hand, a false negative would be the model incorrectly identifying spam mail as a regular email. 
Both scenarios are bad. But a worst scenario would depend on the user. For example, somebody that is more digitally aware would not want a false positive since they are probably better at identifying if there is spam mail in their normal inbox. So they wouldn't want to miss any important emails going to their spam inbox.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*

For this assignment, I followed all the parts in order beginning with Step 1.  I referred back to the Python examples provided from class and the lab to help with the specifics of using train_test_split as well as fitting the model with X,y data. 

I did use generative ai for this example. I used ChatGPT to help me get a better understanding of true and false positives/negatives. I simply asked chat "Can you give an indepth explanation on how true positive, true negative, false positive, and false negative works in data modeling and machine learning". So I did not need to modify any code as I used ChatGPT to help with understanding my results better.

I had challenges in steps 4-5 with getting the code to work in VSCode. For some reason I have to re-call my variables model 1-3 in Step 4 even though it was already done in Step 3. 

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [331]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y
X, y = yellowbrick.datasets.loaders.load_concrete(data_home=None, return_dataset=False)
print(f"Size of X is: {X.size}, Type of X is: {type(X)}")
print(f"Size of y is: {y.size}, Type of y is: {type(y)}")


Size of X is: 8240, Type of X is: <class 'pandas.core.frame.DataFrame'>
Size of y is: 1030, Type of y is: <class 'pandas.core.series.Series'>


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [332]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_value = (X.isnull().sum().sum())+(y.isnull().sum())
print(f"There are {missing_value} missing values in the dataset")


There are 0 missing values in the dataset


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [333]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0
from sklearn.linear_model import LinearRegression
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state= 0)
lr = LinearRegression().fit(X_train, y_train)

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [334]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import mean_squared_error, r2_score

y_train_pred = lr.predict(X_train)
y_val_pred = lr.predict(X_val)

#Mean squared error
mse_train = mean_squared_error(y_train, y_train_pred)
mse_val = mean_squared_error(y_val_pred,y_val)

#R2_score
r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)

### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [335]:
# TO DO: ADD YOUR CODE HERE
results = pd.DataFrame(columns= ["Training accuracy", "Validation accuracy"])
results.loc[0] = mse_train, mse_val
results.loc[1] = r2_train, r2_val
results.index = ['MSE', 'R2']
print(results)

     Training accuracy  Validation accuracy
MSE         111.358439            95.904136
R2            0.610823             0.623414


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

The linear model produced good results, but they were not amazing or terrible.  The model may be more accurate if the R2 score was closer to 1. But above 0.6 is not bad. For MSE, the validation accuracy being lower than our training accuracy is good as that means we are not overfitting our training data.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*

For part 2 of the assignment. I referred back to part 1 of the assignment for certain parts such as assigning X and y from the yellowbrick data set.  I also referred back to the Linear Regression notebook that was covered in class. However, in order to use the built-in sklearn functions (MSE, and R2_score) I used the code from the sklearn website for both MSE and R2_score. (cited below). I did not use any generative AI for this section. The hardest part was figuring out how to use the MSE and R2_score functions and making sense of the results.

sklearn.metrics.mean_squared_error, https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html

sklearn.metrics.r2_score, https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html

Mean Square Error & R2 Score Clearly Explained, https://www.bmc.com/blogs/mean-squared-error-r2-and-variance-in-regression-analysis/#:~:text=There%20is%20no%20correct%20value,means%20the%20model%20is%20perfect.



## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*

From the results, we can see that for the MSE score, our validation model is a bit more accurate compared to our training model. This may be due to some more outliers in the training set.  Since our R2 scores are similar for training and validation, we can say that our model has low variance.  Since our R2 values are around 0.6, we can also say that the model has moderate bias. 

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*

I liked that this assignment built on things that we covered in class. It made it easy to follow along for the steps, but worked well as a review for the things we learnt about in class.  The questions at the end of each part made this assignment more interesting, as we had to explain the answers we got to, not just give our code.

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [336]:
# TO DO: ADD YOUR CODE HERE

#Ridge regression
from sklearn.linear_model import Ridge 
ridge = Ridge().fit(X_train, y_train) 
y_train_pred = ridge.predict(X_train)
y_val_pred = ridge.predict(X_val)

#Mean squared error
mse_train = mean_squared_error(y_train, y_train_pred)
mse_val = mean_squared_error(y_val_pred,y_val)

#R2_score
r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)

ridge_results = pd.DataFrame(columns= ["Training accuracy", "Validation accuracy"])
ridge_results.loc[0] = mse_train, mse_val
ridge_results.loc[1] = r2_train, r2_val
ridge_results.index = ['MSE', 'R2']
print(ridge_results)

     Training accuracy  Validation accuracy
MSE         111.358439            95.904035
R2            0.610823             0.623415


In [337]:
#Lasso regression
from sklearn.linear_model import Lasso
lasso = Lasso().fit(X_train, y_train)
y_train_pred = lasso.predict(X_train)
y_val_pred = lasso.predict(X_val)

#Mean squared error
mse_train = mean_squared_error(y_train, y_train_pred)
mse_val = mean_squared_error(y_val_pred,y_val)

#R2_score
r2_train = r2_score(y_train, y_train_pred)
r2_val = r2_score(y_val, y_val_pred)

lasso_results = pd.DataFrame(columns= ["Training accuracy", "Validation accuracy"])
lasso_results.loc[0] = mse_train, mse_val
lasso_results.loc[1] = r2_train, r2_val
lasso_results.index = ['MSE', 'R2']
print(lasso_results)

     Training accuracy  Validation accuracy
MSE         111.419648            95.584678
R2            0.610609             0.624669


After running the ridge regression, and lasso regression. The accuracy results did not change very much, and changing the alphas did not affect my results very much. The R^2 score at 61% is probably not good enough, as you would typically want to see a higher number.